src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx

```diff
@@ -91,21 +81,95 @@ const DiffExprCompute = (props) => {
+    setSampleKeys(samples.map(sample => sample.key));
+
   }, [hierarchy, properties]);
 
-  const validateForm = () => {
+  const cellIdToSampleMap = useMemo(() => {
+    const mapping = [];
+    sampleKeys.forEach((key, idx) => {
+      const cellIds = properties[key].cellIds;
+      cellIds.forEach(cellId => mapping[cellId] = idx);
+    });
 
-    if (!comparisonGroup[selectedComparison]?.cellSet) {
-      setIsFormValid(false);
-      return;
+    return mapping;
+  }, [numSamples]);
+
+  const canRunDiffExpr = useCallback(() => {
+
+    if (selectedComparison === ComparisonType.WITHIN) return true;
+
+    if (!basis || !cellSet || !compareWith || !cellIdToSampleMap.length > 0) { return false; }
+
+    const basisCellSetKey = getCellSetName(basis);
+
+    const cellSetKey = getCellSetName(cellSet);
+    const compareWithKey = getCellSetName(compareWith);
+
+    let basisCellSet = [];
+    if (basisCellSetKey === 'all') {
+      const allCellIds = sampleKeys.reduce((acc, key) => {
+        const cellIds = properties[key].cellIds;
+        return acc.concat(Array.from(cellIds));
+      }, []);
+      basisCellSet = new Set(allCellIds);
+    } else {
+      basisCellSet = properties[basisCellSetKey].cellIds;
+    }
+
+    const cellSetCellIds = Array.from(properties[cellSetKey].cellIds);
+
+    let compareWithCellIds = [];
+    if (['rest', 'background'].includes(compareWithKey)) {
+      const parentGroup = getRootKey(cellSet);
+
+      const otherGroupKeys = hierarchy.find(obj => obj.key === parentGroup)
+        .children.filter(child => child.key !== cellSetKey);
+
+      compareWithCellIds = otherGroupKeys.reduce((acc, child) => {
+        return acc.concat(Array.from(properties[child.key].cellIds));
+      }, []);
+    } else {
+      compareWithCellIds = Array.from(properties[compareWithKey].cellIds);
     }
 
-    if (!comparisonGroup[selectedComparison]?.compareWith) {
+    // Intersect the basis cell set with each group cell set
+    const filteredGroup1CellIds = cellSetCellIds.filter(cellId => basisCellSet.has(cellId));
+    const filteredGroup2CellIds = compareWithCellIds.filter(cellId => basisCellSet.has(cellId));
+
+    // Prepare an array of length sampleIds to hold tally of cells for each sample in each group
+    const cellSetSamples = new Array(numSamples).fill(0);
+    const compareWithSamples = new Array(numSamples).fill(0);
+
+    // Count the number of cells in each sample
+    filteredGroup1CellIds
+      .forEach(cellId => {
+        const sampleIdx = cellIdToSampleMap[cellId];
+        cellSetSamples[sampleIdx] += 1;
+      });
+
+    filteredGroup2CellIds
+      .forEach(cellId => {
+        const sampleIdx = cellIdToSampleMap[cellId];
+        compareWithSamples[sampleIdx] += 1;
+      });
+
+    const cellSetHasMoreThanMinNumCells = cellSetSamples.find(numCells => numCells > MIN_NUM_CELLS_IN_GROUP);
+    const compareWithHasMoreThanMinNumCells = compareWithSamples.find(numCells => numCells > MIN_NUM_CELLS_IN_GROUP);
```

Comment:
Seems like you do the same thing twice: once for `cellSet` and once for `compareWith`. What you can do is:

Move this code in 1 function called `hasMoreThanMinNumCells` and call it twice. The function will return a boolean.
For each call of the function, you pass in either `compareWithCellIds` or `basisCellSet` as a parameter, depending on whether the call is for `cellSet` or for `compareWith`.

Reply:
Refactored into a function